cmd/find.go

```diff
@@ -0,0 +1,468 @@
+	errorIf(err, "Unable to construct client")
+	fileContent := parseContent(content)
+
+	if noAlias {
+		//pass hostname (alias) to to watchEvents so that it can remove from pathnames returned from watch
+		watchEvents(ctx, clnt, params, alias, fileContent)
+		return
+	}
+	watchEvents(ctx, clnt, params, "", fileContent)
+	return
+}
+
+// doFindOlder checks to see if the given object was created before or after the given time at which an object was created
+func doFindOlder(createTime time.Time, pattern string) bool {
+	i, err := TimeHelper(pattern)
+	now := time.Now()
+	fatalIf(probe.NewError(err), "Error parsing string passed to flag older")
+
+	//find all time in which the time in which the object was just created is after the current time
+	t := time.Date(now.Year(), now.Month(), now.Day()-i, now.Hour(), now.Minute(), 0, 0, time.UTC)
+	return createTime.Before(t)
+}
+
+// doFindNewer checks to see if the given object was created before the given threshold
+func doFindNewer(createTime time.Time, pattern string) bool {
+	i, err := TimeHelper(pattern)
+	now := time.Now()
+
+	fatalIf(probe.NewError(err), "Error parsing string passed to flag newer")
+
+	t := time.Date(now.Year(), now.Month(), now.Day()-i, now.Hour(), now.Minute(), 0, 0, time.UTC)
+	return createTime.After(t) || createTime.Equal(t)
+}
+
+// doFindLargerSize checks to see if the given object is larger than the given threshold
+func doFindLargerSize(size int64, pattern string) bool {
+	i, err := humanize.ParseBytes(pattern)
+	fatalIf(probe.NewError(err), "Error parsing string passed to flag larger")
+
+	return int64(i) < size
+}
+
+// doFindSmallerSize checks to see if the given object is smaller than the given threshold
+func doFindSmallerSize(size int64, pattern string) bool {
+	i, err := humanize.ParseBytes(pattern)
+	fatalIf(probe.NewError(err), "Error parsing string passed to flag smaller")
+
+	return int64(i) > size
+}
+
+// DoFind is used to handle most of the users input
+func DoFind(clnt Client, ctx *cli.Context) {
+	pathnameParts := strings.SplitAfter(ctx.Args().Get(0), "/")
+	alias := strings.TrimSuffix(pathnameParts[0], "/")
+	_, err := getHostConfig(alias)
+
+	// iterate over all content which is within the given directory
+	for content := range clnt.List(true, false, DirNone) {
+		fileContent := parseContent(content)
+		filePath := fileContent.Key
+
+		// traversing in a object store not a file path
+		if err == nil {
+			filePath = path.Join(alias, filePath)
+		}
+
+		if ctx.String("maxdepth") != "" {
+			i, e := strconv.Atoi(ctx.String("maxdepth"))
+			s := ""
+
+			fatalIf(probe.NewError(e), "Error parsing string passed to flag maxdepth")
+
+			// we are going to be parsing the path by x amounts
+			pathParts := strings.SplitAfter(filePath, "/")
+
+			if len(pathParts) < i || i < 0 {
+				i = len(pathParts)
+			}
+
+			// append portions of path into a string
+			for j := 0; j < i; j++ {
+				s += pathParts[j]
+			}
+
+			filePath = s
+			fileContent.Key = s
+		}
```

Comment:
I was trying this out and there appears to be a bug:

```shell
$ mc find myminio/bucket1
myminio/bucket1/NOTICE
myminio/bucket1/a/NOTICE
myminio/bucket1/a/b/NOTICE
myminio/bucket1/a/b/c/NOTICE
$ mc find myminio/bucket1 --maxdepth 1
myminio/
myminio/
myminio/
myminio/
$ mc find myminio/bucket1 --maxdepth 2
myminio/bucket1/
myminio/bucket1/
myminio/bucket1/
myminio/bucket1/
$ mc find myminio/bucket1 --maxdepth 3
myminio/bucket1/NOTICE
myminio/bucket1/a/
myminio/bucket1/a/
myminio/bucket1/a/
$ mc find myminio/bucket1 --maxdepth 4
myminio/bucket1/NOTICE
myminio/bucket1/a/NOTICE
myminio/bucket1/a/b/
myminio/bucket1/a/b/
$ mc find myminio/bucket1 --maxdepth 5
myminio/bucket1/NOTICE
myminio/bucket1/a/NOTICE
myminio/bucket1/a/b/NOTICE
myminio/bucket1/a/b/c/
$ mc find myminio/bucket1 --maxdepth 6
myminio/bucket1/NOTICE
myminio/bucket1/a/NOTICE
myminio/bucket1/a/b/NOTICE
myminio/bucket1/a/b/c/NOTICE
```

`mc find` should not list directories (and certainly not repeatedly).